File: ios/PoseImageAnalysis.swift
import ExpoModulesCore
import ImageIO
import MediaPipeTasksVision
import UIKit

internal struct PoseImageOptions: Record {
  @Field var modelVariant: String = "full"
  @Field var modelPath: String? = nil
  @Field var maxImageDimension: Int = 2048
  @Field var minPoseDetectionConfidence: Double = 0.35
  @Field var minPosePresenceConfidence: Double = 0.35

  func validate() throws {
    guard (256...4096).contains(maxImageDimension) else { throw PoseMediaError.invalidOptions }
    let confidences = [minPoseDetectionConfidence, minPosePresenceConfidence]
    guard confidences.allSatisfy({ $0.isFinite && (0...1).contains($0) }) else {
      throw PoseMediaError.invalidOptions
    }
  }
}

internal enum PoseImageAnalysis {
  static func analyze(_ location: String, options: PoseImageOptions) throws -> [String: Any] {
    try autoreleasepool {
      try options.validate()
      let url = try PoseModel.localURL(location)
      guard let source = CGImageSourceCreateWithURL(url as CFURL, nil) else {
        throw PoseMediaError.invalidImage
      }
      let decoding: [CFString: Any] = [
        kCGImageSourceCreateThumbnailFromImageAlways: true,
        kCGImageSourceCreateThumbnailWithTransform: true,
        kCGImageSourceThumbnailMaxPixelSize: options.maxImageDimension,
        kCGImageSourceShouldCacheImmediately: true,
      ]
      guard let pixels = CGImageSourceCreateThumbnailAtIndex(source, 0, decoding as CFDictionary)
      else { throw PoseMediaError.invalidImage }
      let configuration = PoseLandmarkerOptions()
      configuration.baseOptions.modelAssetPath = try PoseModel.path(
        variant: options.modelVariant, localPath: options.modelPath)
      configuration.baseOptions.delegate = .CPU
      configuration.runningMode = .image
      configuration.numPoses = 1
      configuration.minPoseDetectionConfidence = Float(options.minPoseDetectionConfidence)
      configuration.minPosePresenceConfidence = Float(options.minPosePresenceConfidence)
      let detector = try PoseLandmarker(options: configuration)
      let image = try MPImage(uiImage: UIImage(cgImage: pixels))
      let started = ProcessInfo.processInfo.systemUptime
      let result = try detector.detect(image: image)
      let duration = (ProcessInfo.processInfo.systemUptime - started) * 1000
      var payload = PoseLandmarkPayload.make(result)
      payload["imageSize"] = ["width": pixels.width, "height": pixels.height]
      payload["inferenceDurationMs"] = duration
      payload["model"] = [
        "variant": options.modelVariant, "delegate": "CPU",
        "source": options.modelPath == nil ? "bundled" : "local",
      ]
      return payload
    }
  }
}

File: android/src/main/java/expo/modules/mediapipepose/PoseImageAnalysis.kt
package expo.modules.mediapipepose

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.os.SystemClock
import androidx.exifinterface.media.ExifInterface
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import java.io.File

internal class PoseImageOptions : Record {
    @Field var modelVariant: String = "full"
    @Field var modelPath: String? = null
    @Field var maxImageDimension: Int = 2048
    @Field var minPoseDetectionConfidence: Double = 0.35
    @Field var minPosePresenceConfidence: Double = 0.35

    fun validate() {
        require(maxImageDimension in 256..2048)
        val confidences = listOf(minPoseDetectionConfidence, minPosePresenceConfidence)
        require(confidences.all { it.isFinite() && it in 0.0..1.0 })
    }
}

internal object PoseImageAnalysis {
    fun analyze(context: Context, location: String, options: PoseImageOptions): Map<String, Any> {
        options.validate()
        val pixels = decode(PoseModel.localFile(location), options.maxImageDimension)
        try {
            val configuration =
                PoseLandmarker.PoseLandmarkerOptions.builder()
                    .setBaseOptions(PoseModel.options(options.modelVariant, options.modelPath))
                    .setRunningMode(RunningMode.IMAGE)
                    .setNumPoses(1)
                    .setMinPoseDetectionConfidence(options.minPoseDetectionConfidence.toFloat())
                    .setMinPosePresenceConfidence(options.minPosePresenceConfidence.toFloat())
                    .build()
            PoseLandmarker.createFromOptions(context, configuration).use { detector ->
                val image = BitmapImageBuilder(pixels).build()
                try {
                    val started = SystemClock.elapsedRealtimeNanos()
                    val result = detector.detect(image)
                    val duration = (SystemClock.elapsedRealtimeNanos() - started) / 1_000_000.0
                    return PoseLandmarkPayload.make(result) +
                        mapOf(
                            "imageSize" to
                                mapOf("width" to pixels.width, "height" to pixels.height),
                            "inferenceDurationMs" to duration,
                            "model" to
                                mapOf(
                                    "variant" to options.modelVariant,
                                    "delegate" to "CPU",
                                    "source" to
                                        if (options.modelPath == null) "bundled" else "local",
                                ),
                        )
                } finally {
                    image.close()
                }
            }
        } finally {
            if (!pixels.isRecycled) pixels.recycle()
        }
    }

    private fun decode(file: File, maximumDimension: Int): Bitmap {
        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeFile(file.path, bounds)
        require(bounds.outWidth > 0 && bounds.outHeight > 0) { "Image could not be decoded" }
        require(bounds.outWidth.toLong() * bounds.outHeight <= 16_777_216L) {
            "Source image exceeds 16,777,216 pixels; resize it before analysis."
        }
        val largestDimension = maxOf(bounds.outWidth, bounds.outHeight)
        var sampleSize = 1
        while (largestDimension > maximumDimension.toLong() * sampleSize) sampleSize *= 2
        val decoding =
            BitmapFactory.Options().apply {
                inSampleSize = sampleSize
                inPreferredConfig = Bitmap.Config.ARGB_8888
                inScaled = false
            }
        val pixels =
            requireNotNull(BitmapFactory.decodeFile(file.path, decoding)) {
                "Image could not be decoded"
            }
        try {
            val exif = ExifInterface(file)
            val transform =
                Matrix().apply {
                    // ExifInterface defines rotation after the horizontal flip.
                    if (exif.isFlipped) postScale(-1f, 1f)
                    postRotate(exif.rotationDegrees.toFloat())
                }
            val upright =
                Bitmap.createBitmap(pixels, 0, 0, pixels.width, pixels.height, transform, true)
            if (upright !== pixels) pixels.recycle()
            return upright
        } catch (error: Exception) {
            pixels.recycle()
            throw error
        }
    }
}
